### DISCUSSION TITLE
Is There a Way to Use the Delta Update Feature with a pkg File?

### DISCUSSION THREAD

**Discussion by wogus3602** [sparkle-project/Sparkle#2441] [Q&A] [answered]
Previously, we uploaded a pkg file to the server to download the latest version. This method takes a significant amount of time to update, so we intend to utilize the Delta update feature.

For Delta update, we used generate_appcast on a folder path containing two zipped apps to create delta and xml files.

Herein lies the issue; the two zip files are in .app format, whereas previously, we downloaded through a pkg file. Is there a method to proceed with Delta updates using a pkg file? If not, do we need to upload these two zip files to the server?

### The xml you were using before
```xml
<?xml version="1.0" encoding="utf-8"?>
<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" xmlns:dc="http://purl.org/dc/elements/1.1/" version="2.0">
  <channel>
    <title>12.3.1</title>
    <item>
      <sparkle:version>1470</sparkle:version>
      <sparkle:shortVersionString>12.3.1</sparkle:shortVersionString>
      <pubDate>Wed, 27 Sep 2023 09:20:35 +0900</pubDate>
      <enclosure url="https://.../latest" sparkle:edSignature="" length="413356929" type="application/octet-stream" sparkle:installationType="package"/>
    </item>
  </channel>
</rss>
```

### xml file generated by generate_appcast
```xml
<?xml version="1.0" standalone="yes"?>
<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" version="2.0">
    <channel>
        <title>Elixir</title>
        <item>
            <title>12.3.1</title>
            <pubDate>Wed, 27 Sep 2023 18:42:47 +0900</pubDate>
            <sparkle:version>1475</sparkle:version>
            <sparkle:shortVersionString>12.3.1</sparkle:shortVersionString>
            <sparkle:minimumSystemVersion>12.0</sparkle:minimumSystemVersion>
            <enclosure url="1471.zip" length="414041768" type="application/octet-stream"/>
            <sparkle:deltas>
                <enclosure url="//1475-1470.delta" sparkle:deltaFrom="1470" length="4723918" type="application/octet-stream" sparkle:deltaFromSparkleExecutableSize="1381488" sparkle:deltaFromSparkleLocales="de,he,ar,el,ja,fa,en" sparkle:edSignature="..."/>
            </sparkle:deltas>
        </item>
        <item>
            <title>12.3.1</title>
            <pubDate>Wed, 27 Sep 2023 18:40:13 +0900</pubDate>
            <sparkle:version>1470</sparkle:version>
            <sparkle:shortVersionString>12.3.1</sparkle:shortVersionString>
            <sparkle:minimumSystemVersion>12.0</sparkle:minimumSystemVersion>
            <enclosure url="1470.zip" length="414088680" type="application/octet-stream"/>
        </item>
    </channel>
</rss>
```

**Comment by zorgiepoo** [comment]
Delta updates cannot be used if the new update being served is a pkg, however it looks like you are switching from pkg to regular app distribution (which is a good thing!). The new XML looks OK to me and should work, assuming the .app in 1470.zip and the 1470 one previously produced by your pkg are equivalent. In the new XML, the 1470 entry is not actually necessary; Sparkle will never pick it* since 1475 is greater and they have the same branch point (i.e same minimumSystemVersion, same channel). It doesn't hurt having it, though, and generate_appcast will later manage how many updates are kept in the feed when you start generating more updates.

I assume you left out the sparkle:edSignature from your snippets as those should be there for .zip entries too.

(\*assuming you didn't go out of your way to add any custom logic for selectively picking updates)

**Reply by wogus3602** [reply]
@zorgiepoo 

After checking the delta in the URL of sparkle:deltas, if it's incorrect, do I download the enclosure mentioned earlier?

I updated using the xml below, but the delta doesn't seem to work and appears to download a new file immediately.
I verified that it was the correct delta file using ./Sparkle/BinaryDelta apply.

```xml
<?xml version="1.0" standalone="yes"?>
<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" version="2.0">
    <channel>
        <title>Delta Test</title>
        <item>
            <title>12.3.1</title>
            <pubDate>Tue, 03 Oct 2023 00:05:07 +0900</pubDate>
            <sparkle:version>1485</sparkle:version>
            <sparkle:shortVersionString>12.3.1</sparkle:shortVersionString>
            <sparkle:minimumSystemVersion>12.0</sparkle:minimumSystemVersion>
            <enclosure url="latest link" length="414053165" type="application/octet-stream"/>
            <sparkle:deltas>
                <enclosure url="https://firebasestorage.googleapis.com/v0/b/.com/o/delta.delta" sparkle:deltaFrom="1484" length="167662" type="application/octet-stream" sparkle:deltaFromSparkleExecutableSize="1381488" sparkle:deltaFromSparkleLocales="de,he,ar,el,ja,fa,en" sparkle:edSignature="/..."/>
            </sparkle:deltas>
        </item>
    </channel>
</rss>

```

**Reply by zorgiepoo** [reply]
Please check your app (or Autoupdate but unlikely in this case) for any logs that Sparkle may output -- these would also show up in the Console app. It may log the reason why it skipped the delta update. Experimentally you could also try to omit deltaFromSparkleLocales and deltaFromSparkleExecutableSize as their purpose is for skipping deltas ahead of time when some things in the running app has been stripped by the user.

**Reply by wogus3602** [reply]
I checked the log through console.app.
It seems that the edSignature of the delta file is the problem.
I created appcast and delta files via generate_appcast. I then called checkUpdate and it skipped delta.
Did I miss something?

![스크린샷 2023-10-04 오후 9 31 36](https://github.com/sparkle-project/Sparkle/assets/46857148/80a26b9a-2054-41f5-8d78-7c6a90d90a11)



<img width="722" alt="스크린샷 2023-10-04 오후 9 37 07" src="https://github.com/sparkle-project/Sparkle/assets/46857148/828821ed-6ba3-47fa-b4ca-be094be93df6">

**Reply by zorgiepoo** [reply, marked as answer]
That means it actually did download the delta update but rejected it upon signature verification. One of our tools includes an option to verify a signature so check that. (sign_update I think). Next I would check the delta file your app is downloading from your server is the actual delta file.
